File: shared/time/shanghai_time.js
"use strict";

/**
 * 上海时区时间工具。
 * 这里统一处理交易日、交易时段、日期标准化和截止时间判断。
 */

const SHANGHAI_TIMEZONE = "Asia/Shanghai";

function nowIso() {
  return new Date().toISOString();
}

function getShanghaiParts(date = new Date()) {
  const formatter = new Intl.DateTimeFormat("en-CA", {
    timeZone: SHANGHAI_TIMEZONE,
    hour12: false,
    hourCycle: "h23",
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
    hour: "2-digit",
    minute: "2-digit",
    second: "2-digit",
    weekday: "short",
  });
  const parts = Object.fromEntries(formatter.formatToParts(date).map((item) => [item.type, item.value]));
  const weekdayMap = { Mon: 1, Tue: 2, Wed: 3, Thu: 4, Fri: 5, Sat: 6, Sun: 0 };
  const rawHour = Number(parts.hour || 0);
  // 部分 Node/ICU 运行环境会把上海时间午夜格式化成 24:xx，这里统一归一成 00:xx。
  const normalizedHour = rawHour === 24 ? 0 : rawHour;

  return {
    date: `${parts.year}-${parts.month}-${parts.day}`,
    hour: normalizedHour,
    minute: Number(parts.minute || 0),
    second: Number(parts.second || 0),
    weekday: weekdayMap[parts.weekday] ?? -1,
  };
}

function normalizeDateText(value) {
  const text = String(value || "").trim();
  if (!text) return "";
  const hit = text.match(/\d{4}-\d{2}-\d{2}/);
  return hit ? hit[0] : text.slice(0, 10);
}

function parseTimeToMinutes(timeText) {
  const hit = String(timeText || "").trim().match(/^(\d{2}):(\d{2})$/);
  if (!hit) return null;
  const hour = Number(hit[1]);
  const minute = Number(hit[2]);
  if (!Number.isFinite(hour) || !Number.isFinite(minute)) return null;
  if (hour < 0 || hour > 23 || minute < 0 || minute > 59) return null;
  return hour * 60 + minute;
}

function isTradingWeekday(date = new Date()) {
  const parts = getShanghaiParts(date);
  return parts.weekday >= 1 && parts.weekday <= 5;
}

function isTradingSession(date = new Date()) {
  const parts = getShanghaiParts(date);
  if (parts.weekday < 1 || parts.weekday > 5) return false;
  const minutes = parts.hour * 60 + parts.minute;
  const inWindow = (startHour, startMinute, endHour, endMinute) => {
    const start = startHour * 60 + startMinute;
    const end = endHour * 60 + endMinute;
    return minutes >= start && minutes <= end;
  };

  return (
    inWindow(9, 30, 11, 30) ||
    inWindow(13, 0, 15, 0) ||
    inWindow(9, 30, 12, 0) ||
    inWindow(13, 0, 16, 0)
  );
}

function isAfterCutoff(date = new Date(), cutoffTime = "16:10") {
  const parts = getShanghaiParts(date);
  if (parts.weekday < 1 || parts.weekday > 5) return false;
  const cutoffMinutes = parseTimeToMinutes(cutoffTime);
  if (cutoffMinutes === null) return false;
  return parts.hour * 60 + parts.minute >= cutoffMinutes;
}

module.exports = {
  SHANGHAI_TIMEZONE,
  nowIso,
  getShanghaiParts,
  normalizeDateText,
  parseTimeToMinutes,
  isTradingWeekday,
  isTradingSession,
  isAfterCutoff,
};

File: shared/time/shanghai_time.py
"""上海时区时间工具。"""

from __future__ import annotations

import re
from datetime import datetime, timezone, timedelta
from typing import Any, Dict


SHANGHAI_TIMEZONE = timezone(timedelta(hours=8))


def now_iso() -> str:
    """返回当前 UTC ISO 时间。"""

    return datetime.now().isoformat()


def shanghai_now() -> datetime:
    """返回上海时区当前时间。"""

    return datetime.now(SHANGHAI_TIMEZONE)


def get_shanghai_parts(dt: datetime | None = None) -> Dict[str, Any]:
    """把shanghai_time拆成统一的日期与时分秒结构。"""

    current = dt.astimezone(SHANGHAI_TIMEZONE) if dt else shanghai_now()
    weekday = current.weekday() + 1
    if weekday == 7:
        weekday = 0
    return {
        "date": current.strftime("%Y-%m-%d"),
        "hour": current.hour,
        "minute": current.minute,
        "second": current.second,
        "weekday": weekday,
    }


def normalize_date_text(value: Any) -> str:
    """提取 YYYY-MM-DD 格式，兼容含时间的字符串。"""

    text = str(value or "").strip()
    if not text:
        return ""
    hit = re.search(r"\d{4}-\d{2}-\d{2}", text)
    return hit.group(0) if hit else text[:10]


def parse_time_to_minutes(time_text: str | None) -> int | None:
    """把 HH:MM 文本转换成分钟数。"""

    hit = re.match(r"^(\d{2}):(\d{2})$", str(time_text or "").strip())
    if not hit:
        return None
    hour = int(hit.group(1))
    minute = int(hit.group(2))
    if hour < 0 or hour > 23 or minute < 0 or minute > 59:
        return None
    return hour * 60 + minute


def is_trading_weekday(dt: datetime | None = None) -> bool:
    """判断是否为交易日周一到周五。"""

    weekday = get_shanghai_parts(dt)["weekday"]
    return 1 <= weekday <= 5


def is_trading_session(dt: datetime | None = None) -> bool:
    """判断当前是否处于 A/H 常用盘中时段。"""

    parts = get_shanghai_parts(dt)
    if parts["weekday"] < 1 or parts["weekday"] > 5:
        return False
    minutes = parts["hour"] * 60 + parts["minute"]
    return (
        9 * 60 + 30 <= minutes <= 11 * 60 + 30
        or 13 * 60 <= minutes <= 15 * 60
        or 9 * 60 + 30 <= minutes <= 12 * 60
        or 13 * 60 <= minutes <= 16 * 60
    )


def is_after_cutoff(dt: datetime | None = None, cutoff_time: str = "16:10") -> bool:
    """判断当前是否已过某个上海时区截止时间。"""

    parts = get_shanghai_parts(dt)
    if parts["weekday"] < 1 or parts["weekday"] > 5:
        return False
    cutoff_minutes = parse_time_to_minutes(cutoff_time)
    if cutoff_minutes is None:
        return False
    return parts["hour"] * 60 + parts["minute"] >= cutoff_minutes

